File: src/euler5.rs
use std::ops::Range;

fn lcm(numbers: Range<u32>) -> u32 {
	let mut iter = numbers.into_iter();
	let mut lcm = iter.next().unwrap();
	for number in iter {
		lcm = lcm_pair(lcm, number);
	}
	lcm
}

fn lcm_pair(a: u32, b: u32) -> u32 {
	(a * b) / gcd_pair(a, b)
}

fn gcd_pair(mut a: u32, mut b: u32) -> u32 {
	loop {
		if b == 0 {
			break;
		}
		let temp = b;
		b = a % b;
		a = temp;
	}
	a
}

/// What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
fn main() {
	println!("{}", lcm(1..20));
}

/// 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
#[test]
fn test() {
	let expected = 2520;
	let result = lcm(1..10);
	assert_eq!(result, expected);
}

File: c/src/euler5.c
//prepare: -lm
#include <stdio.h>
#include <assert.h>
#include "number_theory.c"

/*
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
*/

void swap(long unsigned* a, long unsigned* b){
	long unsigned temp = *b;
	*b = *a;
	*a = temp;
}

long unsigned lcm_one_to(unsigned limit){
	long unsigned answer = 1;
	for (unsigned i = 2; i <= limit; i++) {
		answer = lcml(answer, i);
	}
	return answer;
}

int main(){
	assert(lcm_one_to(10) == 2520);
	printf("%lu\n", lcm_one_to(20));
	return 0;
}